src/main/ApplicationImpl.h

```diff
@@ -120,14 +120,17 @@ class ApplicationImpl : public Application
 
     virtual void reportInfo() override;
 
+    virtual std::shared_ptr<BasicWork>
+    scheduleSelfCheck(bool waitUntilNextCheckpoint) override;
+
     virtual Hash const& getNetworkID() const override;
 
     virtual AbstractLedgerTxnParent& getLedgerTxnRoot() override;
 
     virtual void resetDBForInMemoryMode() override;
 
   protected:
-    std::unique_ptr<LedgerManager>
+    std::shared_ptr<LedgerManager>
```

Comment:
this change is not ideal: Application should be the sole owner of LM, and this allows other parts of the code reference LM, potentially introducing lifetime issues. A better change would be to allow the condition in ConditionalWork to take Application as an argument, as each Work has access to the app, so we can query whatever manager we need in the lambda. This would be similar to what WorkWithCallback is doing (note: we can address this in a subsequent PR as we're trying to wrap up with this one)

Reply:
It's a fair point and something I thought about doing in this change instead. I agree your suggestion and will do it here.